Vol surface implementation

sefhamada opened the pull request:
Fixes #.
 
- Added  "VolSurface" class and an interpolation method (with its generic) for term/strike type of vol.
- The main changes are in the file : vol_surface_class.R
- I added documentations and examples similar to the existing classes in fmbasics
- I added the test harness in testthat folder: test_volsurface.R
- I  performed devtool::check() with 0 error, 0 warnings and 0 notes.

@imanuelcostigan

Commit:
Create testing_debug.R

Commit:
VolSurface class and method

Commit:
vol surface unit testing implemented

Commit:
removing extra script

Commit:
updated the zero curve to match the correct version

Commit:
package checks and ready for pull request

Commit:
package checks

imanuelcostigan commented on R/utils.R:
```diff
@@ -30,4 +30,53 @@ is_atomic_list <- function(ll, .p) {
+#' @family build object helpers
+
+build_vol_surface <- function(){
+##1.651555326
```
Probably something to remove?

sefhamada replied:
done

imanuelcostigan commented on R/generic-methods.R:
```diff
@@ -125,4 +125,21 @@ interpolate_dfs <- function(x, from, to, ...) UseMethod("interpolate_dfs")
+#' implied volatilities at each couple of `(date, strike)`.
+#' @export
+#' @aliases interpolate_vol
+interpolate_vol <- function(x, maturity, strike, ...) UseMethod("interpolate_vol")
```
Why can't you extend `interpolate()` by implementing a specific method for `VolSurface`?

sefhamada replied:
Good idea ! It is now inherited by VolSurface class. I previously thought to follow the same logic as interpolate_dfs, interpolate_zeros, interpolate_fwd; but this is actually different.

imanuelcostigan commented on R/utils.R:
```diff
@@ -30,4 +30,53 @@ is_atomic_list <- function(ll, .p) {
+  spot <- 97.62
+  reference_date <- as.Date("2019-04-26")
+  tenors <- lubridate::ymd(vols$Date)
+  strikes <- seq(from = 0.05, to = 2.55, by= 0.05)*spot
```
Why do you need to hard code this - isn't this in `volsurface.csv"?

sefhamada replied:
The data in "volsurface.csv" does not include this information (ref date, tenors, strike) explicitly. It only contains the tenor dates in the form yyyymmdd, the strikes are given via moneyness: X5., X10., X15., ...., X250., where the digits refer to (K/S)*100. This is an example of a vol surface extracted from Murex.

imanuelcostigan replied:
I mean the first line of the `volsurfaces.csv` file has the percentages you have encoded in `strikes`.

sefhamada replied:
![image](https://user-images.githubusercontent.com/39390900/62014308-96aa9e00-b1e2-11e9-9067-d56557263c1d.png)

 This is the format of  "vols" data sourced from the exdata (volsurface.csv). I now use the first line (colnames) containing the percentage information to retrieve the moneyness and therefore the strikes.

imanuelcostigan commented on R/utils.R:
```diff
@@ -30,4 +30,53 @@ is_atomic_list <- function(ll, .p) {
+  vol_quotes <- vols[,-c(1:4) ]
+  vol_quotes <- t(vol_quotes)/100
+  vol_interp <- LinearTimeVarInterpolation()
+  vol_obj <- VolSurface(reference_date, "strike/term", vol_quotes, strikes, tenors,
```
You don't need to pass the result to `vol_obj` and then return it on the next line. Leave the call to `VolSurface()` as the last line in the function.

sefhamada replied:
That's true, I removed the extra variable assignment. It is a cautious habit.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+#' VolSurface class
+#'
+#' The `VolSurface`` class is designed to represent volatility surfaces.
```
you have one too many back-ticks here.

sefhamada replied:
done

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+#' Checks whether the dimentions of the surface correspond to the right type of axis: moneyness, delta or strike
+#'
+#' @param reference_date `Date` value that represents the as of date
+#' @param surface_type type of the vol dimentions. Available types: "strike/term"
```
spelling mistake here.

sefhamada replied:
fixed dimensions spelling in line 4 and 7.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+#'
+#' @param reference_date `Date` value that represents the as of date
+#' @param surface_type type of the vol dimentions. Available types: "strike/term"
+#' @param vol_quotes `matrix` of class `numeric` holding the vol values.
```
we shouldnt be using matrices as an input type - they are terrible to work with. Use a (tidy) tibble. Suggest the tibble structure is something like:

```csv
maturity, strike, value
20200201, 72.00, 0.20
20200301, 72.00, 0.24
..., ..., ...
20200201, 82.00, 0.25
20200301, 82.00, 0.26
..., ..., ...
```

In the doc be very clear about the structure of this including for example that:

1. The name of the second column (e.g. `strike` or `moneyness` or `delta`) will define what sort of surface you are getting. Be clear about the options that are valid.
2. The string values / formats required for each field: e.g. `maturity` must be a string of format `YYYYMMDD`, `value` much be in decimals...etc.

sefhamada replied:
I changed the structure of VolSurface following the above suggestion. Now it only includes: reference_date, vol_quotes(tibble as described in the example above), ticker and Interpolation.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+                      day_basis = "act/365")
+
+  time_variance <- sweep(variance, MARGIN = 2, tenors_yf, `*`)
+  #time_variance <- vols
```
remove comments like this.

sefhamada replied:
done

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+#' @param interpolation object of type interpolation [fmbasics::Interpolation()]
+#' @export
+
+VolSurface <- function(reference_date, surface_type, vol_quotes, dim1, dim2,
```
Once you are much clearer on what `vol_quotes` should be, you will be able to simplify this interface considerably.

sefhamada replied:
Now the constructor has 4 members, "reference_date", "vol_quotes", "ticker", "interpolation".

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+new_VolSurface <- function(reference_date, surface_type, vol_quotes, dim1, dim2,
+                           ticker, interpolation){
+
+  assertthat::assert_that(is.Interpolation(interpolation))
```
why are you including an assertion in a `new_X()` method? Shouldn't this what the `validate_X()` function is there for?

sefhamada replied:
Here I am using the same logic used in ZeroCurve class where the assertion of the interpolation method is introduced in new_ZeroCurve rather than in validate_ZeroCurve. We can easily shift the assertion clause to the validation function.

imanuelcostigan replied:
The reason for it in ZeroCurve land is because I wanted to make sure when an instance of this was created, it would have a functioning and calibrated interpolation `interpolator`.

sefhamada replied:
Now the assertion on the Interpolation is included in the validation function.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+  assertthat::assert_that(
+                assertthat::is.date(x$reference_date),
+                x$surface_type %in% c("strike/term","delta/term","moneyness/term"),
+                all.equal(dim(x$vol_quotes), c(length(x$dim1), length(x$dim2))),
```
Once you are clearer on `vol_type` structure per above, you will be able to update these assertions.

sefhamada replied:
done

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+interpolate_vol.VolSurface <- function(x, maturity, strike, ...){
+
+  assertthat::assert_that(
+    is.VolSurface(x),
```
Not required. This is method will only be called when `x` is a `VolSurface`

sefhamada replied:
True. Now removed.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+
+#'@rdname interpolate_vol
+#'@export
+interpolate_vol.VolSurface <- function(x, maturity, strike, ...){
```
Can you please use the same language as we use for `interpolate.ZeroCurve()`? THe interface should look like 

```R
function(x, at, ...)
```

where `at` is a named list of length two where each list member is of the same length as the other and can be a vector of terms and strikes (e.g. `list(term = c(1, 2), smile = c(72, 92))`).

sefhamada replied:
Good point ! now the interpolation for VolSurface class inherits from the generic interpolation method, and it is using "at" as per the suggested example. Although I assumed the names of the list members are "term" and "strike" instead of "smile".

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+    #                                    yleft = smile[1], yright = tail(smile,1))
+    #   imp_vol[i] <- smile_interp(tt_yf)
+    #
+    # }
```
Why are these comments in here?

imanuelcostigan replied:
I think this method needs a bit of work to address other comments above. Once this is done, I will review the rest of this.

sefhamada replied:
Please feel free to suggest any improvement to the method. It does look slightly different now that the above comments are incorporated.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,160 @@
+  imp_vol <- rep(NA, length(maturity))
+
+
+  if(is.LinearTimeVarInterpolation(x$interpolation) & surface_type == "strike/term"){
```
Is this if statement necessary? Why isn't this necessary for the `ZeroCurve` class?

sefhamada replied:
I used this "If" statement as the interpolation technique depends on the type of the surface. At this stage we are only accounting for "strike/term" type of surface, and left a place-holder for interpolating other types (delta/term, moneyness/term).

sefhamada replied:
I am aware there are many ways of interpolating Volatilities. I used the "if" statement as the implemented approach is only valid when the interpolation is of type (linear on term, cubic spline on strike) over a time-variance surface, and the surface is of type term/strike.

imanuelcostigan commented on R/zero-curve-class.R:
```diff
@@ -190,7 +190,8 @@ Interpolation <- function(method, what) {
     linear_zeros = "Linear",
-    natural_cubic_zeros = "Cubic"
+    natural_cubic_zeros = "Cubic",
+    linear_time_variance = "Linear"
```
Do you really need a new interpolation method called `LinearTimeVarInterpolation`? Isn't this simply LinearInterpolation applied to time weighted variance?

sefhamada replied:
Even though the name says Linear, but the actual interpolation is Linear on time, Cubic spline on strike. Including the whole description will blow up the length of the method name.  The actual interpolation is two-fold: 1st interpolation on the time, and a 2nd interpolation on the strikes.

imanuelcostigan commented on inst/extdata/volsurface.csv:
```diff
@@ -0,0 +1,18 @@
+Code,Type,Tenor,Date,5%,10%,15%,20%,25%,30%,35%,40%,45%,50%,55%,60%,65%,70%,75%,80%,85%,90%,95%,100%,105%,110%,115%,120%,125%,130%,135%,140%,145%,150%,155%,160%,165%,170%,175%,180%,185%,190%,195%,200%,205%,210%,215%,220%,225%,230%,235%,240%,245%,250%,255%
```
You may need to update the format of this data (or write another function to transform this data into the format suggested in one of my earlier comments)

sefhamada replied:
This data is extracted from a Murex report as an example of vol surface data, it is then massaged in R/utils.R to construct an object following the suggested class structure of VolSurface.

imanuelcostigan commented on inst/extdata/zerocurve.csv:
```diff
@@ -1,28 +1,28 @@
 start,end,zeros,dfs
```
why has this file changed?

sefhamada replied:
Not totally sure, but we had to change the  data in the zero curve to fix the bug in the discount factor interpolation.

imanuelcostigan replied:
You need to explain why this changed (didn't it work previously?).

sefhamada replied:
What happened here is: The data in zerocurve.csv was wrong (wrong pillar points) which caused the zero interpolation of build_zero_curve() to return wrong results. So we have fixed the bug in the commit  "sefhamada:fixing-the-zero-curve" by changing the data in the file. This was done in the master branch. So I have changed the zerocurve.csv data in this fork to match the data in the master.

imanuelcostigan commented on tests/testthat/test-volsurface.R:
```diff
@@ -0,0 +1,19 @@
+  testthat::expect_equal(interpolate_vol(vs, as.Date("2020-02-29"), 150),
+           0.19909034016558932)
+})
+
```
This needs to be slightly more comprehensive. Take your queue from the `test-interpolate.R` file

sefhamada replied:
If I didn't misunderstood the comment, I have added 2 more interpolation points.

imanuelcostigan replied:
yes, but can you also provide some interpolation points that include the region before the first strike and after the last strike / maturity date?

sefhamada replied:
Added a point with strike = 3, on the left of the interpolation support; and a point with strike = 300 on the right of the support with maturity = "2025-06-10" beyond the maturities of the surface data.

imanuelcostigan reviewed (changes requested):
@sefhamada 

- [ ] Please format all files using `styler::style_file()`. 
- [ ] Please reflected on my in-line comments and update your PR. 

Thanks.

Commit:
The package checks after Imanuel comments

Commit:
added the print fct and example for VolSurface interpolation

sefhamada commented:
Just added an interpolation example and print/format for VolSurface objects.

imanuelcostigan commented on R/utils.R:
```diff
@@ -30,4 +30,56 @@ is_atomic_list <- function(ll, .p) {
+
+
+
+
```
You don't need this many new lines at the end of a file.

sefhamada replied:
Blank lines now removed.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+validate_VolSurface <- function(x) {
+  assertthat::assert_that(
+    assertthat::is.date(x$reference_date),
+    colnames(x$vol_quotes)[1] == "maturity",
```
I don't think `vol_quotes` should be defined in such a way that the column order matters.

imanuelcostigan replied:
Should you create a new class `VolQuotes` (and possibly an extension `EquityVolQuotes`) that extends `tibble`? That way you can move these assertions out of VolSurface and into VolQuotes (or extension thereof).

sefhamada replied:
Now created a VolQuotes class that contains the volatility information, for Equity there is no need to extend, however, this class could be inherited for other markets - e.g. interest rates where the tenors dimension is also needed.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    colnames(x$vol_quotes)[1] == "maturity",
+    colnames(x$vol_quotes)[2] %in% c("strike", "delta", "moneyness"),
+    colnames(x$vol_quotes)[3] == "value",
+    tibble::is_tibble(x$vol_quotes),
```
If you move this up a few assertions, you can ensure that `colnames()` calls make sense.

sefhamada replied:
The class of VolQuotes is now designed to to accommodate different types of smiles (strike, delta, moneyness). The user can specify the type as a controlled input.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    colnames(x$vol_quotes)[3] == "value",
+    tibble::is_tibble(x$vol_quotes),
+    is.character(x$vol_quotes$maturity),
+    assertthat::is.date(as.Date(x$vol_quotes$maturity, format = "%Y%m%d")),
```
This seems like an anti-pattern. You probably want to ensure that date fields in vol_quotes are actually dates or date-times. Leave it to the user to make sure their inputs are correct rather than giving them the flexibility (and hence introducing greater surface for bugs to crawl out of).

sefhamada replied:
now the class VolQuotes insures that the maturities are of class date.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    tibble::is_tibble(x$vol_quotes),
+    is.character(x$vol_quotes$maturity),
+    assertthat::is.date(as.Date(x$vol_quotes$maturity, format = "%Y%m%d")),
+    is.double(x$vol_quotes$strike),
```
Why not `is.numeric()`?

sefhamada replied:
I would have used numeric, I am using the suggestions from the comment made above being decimal:
![image](https://user-images.githubusercontent.com/39390900/62341723-acdc9500-b527-11e9-9caf-a5fb5718163e.png)

sefhamada replied:
We want the vols to be in decimals and not in percentage, so it is just using numeric.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    is.double(x$vol_quotes$strike),
+    is.double(x$vol_quotes$value),
+    !is.unsorted(unique(x$vol_quotes$strike), na.rm = F),
+    !is.unsorted(as.Date(unique(x$vol_quotes$maturity), format = "%Y%m%d"), na.rm = F),
```
This is another example of where your anti-pattern above reemerges.

sefhamada replied:
Now these assertions are moved to VolQuotes object validation; and the user has to provide the data with sorted input dimensions similar to ZeroCurve, otherwise it would be messy.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    assertthat::is.date(as.Date(x$vol_quotes$maturity, format = "%Y%m%d")),
+    is.double(x$vol_quotes$strike),
+    is.double(x$vol_quotes$value),
+    !is.unsorted(unique(x$vol_quotes$strike), na.rm = F),
```
Use `FALSE` instead of `F` and `TRUE` instead of `T`.

sefhamada replied:
done

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    length(x$vol_quotes$strike) == length(unique(x$vol_quotes$strike)) *
+      length(unique(x$vol_quotes$maturity)),
+    is.character(x$ticker),
+    is.Interpolation(x$interpolation)
```
You have already checked for this in `new_VolSurface()`

sefhamada replied:
Now all the assertions are moved to validate_VolSurface()

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+
+#' @export
+print.VolSurface  <- function(x, ...){
+
```
No line spacing required here

sefhamada replied:
now removed.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+
+  cat(format(x), "\n")
+  print(x$vol_quotes)
+
```
Or here

sefhamada replied:
Removed

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+#' list(term = c(1, 2), strike = c(72, 92)).
+#' @param ... unused in this model.
+#' @return `numeric` vector with length equal to the length of `at` members.
+#' @examples vol_surface <- fmbasics::build_vol_surface()
```
You don't need to qualify the package namespace in examples contained in the same package

sefhamada replied:
Very true, now removed.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+      imp_vol[i] <- sqrt(interpolated_time_var / tt_yf)
+    }
+  } else {
+    print("Interpolation not implemented for this surface type")
```
Not necessary because `new_VolSurface()` doesn't permit other options

sefhamada replied:
The logic of interpolation is now changed so that only accept the pre-implemented methods.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+    is.numeric(at$strike)
+  )
+
+  if (is.LinearTimeVarInterpolation(x$interpolation) & colnames(x$vol_quotes)[2] == "strike") {
```
Is there a better way of handling situations where quotes are not according to strike?

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+      for (k in 1:length(smile)) {
+        g <- stats::approxfun(
+          x = unique(tenors_yf),
+          y = time_variance[(1 + (k - 1) * length(unique(tenors_yf))):(k * length(unique(tenors_yf)))],
```
This doesn't seem right. the length of your `x` values can be different to the length of your `y` values. What is the behaviour of `stats::approxfun()` in this case?

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+        interpolated_time_var <- utils::tail(smile, 1)
+      }
+      imp_vol[i] <- sqrt(interpolated_time_var / tt_yf)
+    }
```
I think it may be more efficient (and consistent with `ZeroCurve`) if the calibration interpolator function is stored in the instance of `VolSurface`. Some part of this shouldn't change for a given `(term, strike)` that you are looking to interpolate.

sefhamada replied:
This now is the case where the interpolator is embedded in the VolSurface object - calling a separate mathematical Linear_Cubic implemented in utils.R file. Similar to ZeroCurve class logic.

imanuelcostigan commented on R/vol-surface-class.R:
```diff
@@ -0,0 +1,173 @@
+      )
+      for (k in 1:length(smile)) {
+        g <- stats::approxfun(
+          x = unique(tenors_yf),
```
This seems to be the way you are trying to deal with `vol_quotes` being a tidy tibble. See next comment - is this really right?

sefhamada replied:
The main conditions for the vols data are sorted dimensions and for the data to be a rectangular data.

imanuelcostigan reviewed (changes requested):
@sefhamada See comments below. There are some further things to correct / change and I still have a few qs. I also think I have overestimated styler's abilities out of the box. There is some code style that isn't quite right. I'll fix this up once we merge this PR into master.

Commit:
included the comments in  v0.4.0 milestone. package checked for PR

Commit:
Removing a conflict

In the previous commit I have used "@family build object helpers" for the vol objects, this caused conflict in the Rd file, I had to remove it . Now created a new family for vol object build. We can re-link the families a-posteriori

Commit:
Documentation typo

imanuelcostigan reviewed (approved):
I will merge this PR into a branch on `fmbasics` and do a bit more work on it before I merge into master. I'll tag you when I'm done.
